fdbclient/NativeAPI.actor.cpp

```diff
@@ -649,17 +653,23 @@ ACTOR Future<Void> updateCachedRanges(DatabaseContext* self, std::map<UID, Stora
 	}
 }
 
+// The reason for getting a pointer to DatabaseContext instead of a reference counted object is because reference
+// counting will increment reference count for DatabaseContext which holds the future of this actor. This creates a
+// cyclic reference and hence this actor and Database object will not be destroyed at all.
 ACTOR Future<Void> monitorCacheList(DatabaseContext* self) {
-	state Database db(self);
-	state Transaction tr(db);
+	state Transaction tr;
 	state std::map<UID, StorageServerInterface> cacheServerMap;
 	state Future<Void> updateRanges = updateCachedRanges(self, &cacheServerMap);
 	// if no caches are configured, we don't want to run this actor at all
 	// so we just wait for the first trigger from a storage server
 	wait(self->updateCache.onTrigger());
 	try {
 		loop {
-			tr.reset();
+			// Need to make sure that we eventually destroy tr. We can't rely on getting cancelled to do this because of
+			// the cyclic reference to self.
+			tr = Transaction();
+			wait(delay(0)); // Give ourselves the chance to get cancelled if self was destroyed
+			tr = Transaction(Database(Reference<DatabaseContext>::addRef(self)));
```

Comment:
This looks like a good solution. It might be easier to understand the flow here if this was replaced with a function:
```
// Need to make sure that we eventually destroy tr. We can't rely on getting cancelled to do this because of
// the cyclic reference to self.
ACTOR static Future<Void> refreshTransaction(DatabaseContext *self, Transaction *tr) {
  *tr = Transaction();
  wait(delay(0)); // Give ourselves the chance to get cancelled if self was destroyed
  *tr = Transaction(Database(Reference<DatabaseContext>::addRef(self));
}
```
that can be used here and in `clientStatusUpdateActor`.

Reply:
Done